packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,280 @@
+ */
+function getNamespacedStorageItems(
+  node: StructDefinition,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+): StorageItem[] {
+  const astTypeMembers = getTypeMembers(node, { typeName: true });
+
+  const storageItems: StorageItem[] = [];
+
+  for (const astMember of astTypeMembers) {
+    let item: StorageItem = {
+      contract: context.contractDef.name,
+      label: astMember.label,
+      type: astMember.type,
+      src: decodeSrc({ src: getOriginalMemberSrc(node.canonicalName, astMember.label, origContractDef) }),
+    };
+
+    const layoutMember = findLayoutStructMember(
+      { ...context.storageLayout?.types },
+      node.canonicalName,
+      astMember.label,
+    );
+
+    if (layoutMember?.offset !== undefined && layoutMember?.slot !== undefined) {
+      item = {
+        ...item,
+        offset: layoutMember.offset,
+        slot: layoutMember.slot,
+      };
+    }
+
+    storageItems.push(item);
+
+    loadLayoutType(astMember.typeName, layout, context.deref);
+  }
+  return storageItems;
+}
+
+/**
+ * Gets the struct definition matching the given canonical name from the original contract definition.
+ */
+function getOriginalStruct(structCanonicalName: string, origContractDef: ContractDefinition) {
+  for (const node of origContractDef.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      if (node.canonicalName === structCanonicalName) {
+        return node;
+      }
+    }
+  }
+  throw new Error(
+    `Could not find original source location for namespace struct with name ${structCanonicalName} from contract ${origContractDef.name}`,
+  );
+}
+
+/**
+ * Gets the original source location for the given struct canonical name and struct member label.
+ */
+function getOriginalMemberSrc(structCanonicalName: string, memberLabel: string, origContractDef: ContractDefinition) {
+  const node = getOriginalStruct(structCanonicalName, origContractDef);
+  if (node !== undefined) {
+    const typeMembers = getTypeMembers(node, { src: true });
+
+    for (const member of typeMembers) {
+      if (member.label === memberLabel) {
+        return member.src;
+      }
+    }
+  }
+
+  throw new Error(
+    `Could not find original source location for namespace struct with name ${structCanonicalName} and member ${memberLabel}`,
+  );
+}
+
+/**
+ * From the given layout types, gets the struct member matching the given struct canonical name and struct member label.
+ */
+function findLayoutStructMember(
+  types: Record<string, TypeItem<string>>,
+  structCanonicalName: string,
+  memberLabel: string,
+) {
+  const structType = findTypeWithLabel(types, `struct ${structCanonicalName}`);
+  const structMembers = structType?.members;
+  if (structMembers !== undefined) {
+    for (const structMember of structMembers) {
+      assert(typeof structMember !== 'string');
```

Comment:
Needs import `isStructMembers`.

```suggestion
    assert(isStructMembers(structMembers));
    for (const structMember of structMembers) {
```